skfda/preprocessing/dim_reduction/projection/_fpca.py

```diff
@@ -1,33 +1,433 @@
 
-    Performs Functional Principal Components Analysis to reduce
-    dimensionality and obtain the principal modes of variation for a
-    functional data object.
+    Attributes:
+        n_components (int): number of principal components to obtain from
+            functional principal component analysis. Defaults to 3.
+        centering (bool): if True then calculate the mean of the functional data
+            object and center the data first
+    """
+
+    def __init__(self, n_components=3, centering=True):
+        """FPCA constructor
+
+        Args:
+            n_components (int): number of principal components to obtain from
+                functional principal component analysis
+            centering (bool): if True then calculate the mean of the functional
+                data object and center the data first. Defaults to True
+        """
+        self.n_components = n_components
+        self.centering = centering
+
+    @abstractmethod
+    def fit(self, X, y=None):
+        """Computes the n_components first principal components and saves them
+        inside the FPCA object.
+
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present for convention of a fit function
+
+        Returns:
+            self (object)
+        """
+        pass
+
+    @abstractmethod
+    def transform(self, X, y=None):
+        """Computes the n_components first principal components score and
+        returns them.
+
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present because of fit function convention
+
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+        pass
 
-    It uses SVD numpy implementation to compute PCA.
+    def fit_transform(self, X, y=None, **fit_params):
+        """Computes the n_components first principal components and their scores
+        and returns them.
+        Args:
+            X (FDataGrid or FDataBasis):
+                the functional data object to be analysed
+            y (None, not used):
+                only present for convention of a fit function
 
-    Args:
-        fdatagrid (FDataGrid): functional data object.
-        n (int, optional): Number of principal components. Defaults to 2.
+        Returns:
+            (array_like): the scores of the data with reference to the
+            principal components
+        """
+        self.fit(X, y)
+        return self.transform(X, y)
 
-    Returns:
-        tuple: (scores, principal directions, eigenvalues)
+
+class FPCABasis(FPCA):
+    """Funcional principal component analysis for functional data represented
+    in basis form.
+
+    Attributes:
+        n_components (int): number of principal components to obtain from
```

Comment:
The attributes set in `__init__` should be parameters, not attributes.

Reply:
Solved